lib/HLSL/HLOperationLower.cpp

```diff
@@ -2064,25 +2064,26 @@ Value *TrivialDotOperation(OP::OpCode opcode, Value *src0,
   return dotOP;
 }
 
-Value *TranslateIDot(Value *arg0, Value *arg1, unsigned vecSize, hlsl::OP *hlslOP, IRBuilder<> &Builder) {
+Value *TranslateIDot(Value *arg0, Value *arg1, unsigned vecSize, hlsl::OP *hlslOP, IRBuilder<> &Builder, bool isSigned = true) {
+  auto madOpCode = isSigned ? DXIL::OpCode::IMad : DXIL::OpCode::UMad;
   Value *Elt0 = Builder.CreateExtractElement(arg0, (uint64_t)0);
   Value *Elt1 = Builder.CreateExtractElement(arg1, (uint64_t)0);
   Value *Result = Builder.CreateMul(Elt0, Elt1);
   switch (vecSize) {
   case 4:
     Elt0 = Builder.CreateExtractElement(arg0, 3);
     Elt1 = Builder.CreateExtractElement(arg1, 3);
-    Result = TrivialDxilTrinaryOperation(DXIL::OpCode::IMad, Elt0, Elt1, Result, hlslOP, Builder);
+    Result = TrivialDxilTrinaryOperation(madOpCode, Elt0, Elt1, Result, hlslOP, Builder);
```

Comment:
This isn't due to your change, but this switch seems an odd way to generate the MADs, even though it's numerically equivalent.  I think I'd prefer a simple loop here, like:
```
for (unsigned iVecElt = 1; iVecElt < vecSize; ++iVecElt) {
  Elt0 = Builder.CreateExtractElement(arg0, iVecElt);
  Elt1 = Builder.CreateExtractElement(arg1, iVecElt);
  Result = TrivialDxilTrinaryOperation(madOpCode, Elt0, Elt1, Result, hlslOP, Builder);
}
```
That way, ordering of extractelements and uses are nice and sequential.

Reply:
updated.